std/functional.d

```diff
@@ -1129,8 +1129,9 @@ template memoize(alias fun)
     {
         alias Args = Parameters!fun;
         import std.typecons : Tuple;
+        import std.traits : Unqual;
 
-        static ReturnType!fun[Tuple!Args] memo;
+        static Unqual!(ReturnType!fun)[Tuple!Args] memo;
```

Comment:
The `Unqual` really shouldn't be necessary (in a perfect world), but I guess the compiler can't recognize that.

Reply:
Given `const(int) fun(int a, int b);`, `typeof(memoize!fun.memo)` becomes: `ReturnType!fun[Tuple!Args]` -> `const(int)!fun[Tuple!(int, int)]`. Later on, for the line `memo[t] = fun(args)` you get:
```
Error: cannot modify const expression memo[t]
```

Reply:
Yep, but since this is insert-only, it should work.
However, it would take the compiler to recognize insert-only path (e.g. by handling `require` ?) to make this work.

Reply:
I know, I thought as well and I even tried using `require` and without `Unqual` change, but it has this [line](https://github.com/dlang/druntime/blob/master/src/object.d#L2600) inside, which would fail because `*p` is `const(int)`.

Reply:
That's said, I think it's a would be a good enhancement request.

Reply:
`Unqual + emplace` should work there, but I think that change is better left for the master branch, while this PR is targeting `stable`.